library/gcm.c

```diff
@@ -880,6 +880,13 @@ int mbedtls_gcm_self_test(int verbose)
             mbedtls_printf("  GCM note: using AESNI.\n");
         } else
 #endif
+
+#if defined(MBEDTLS_AESCE_C) && defined(MBEDTLS_HAVE_ARM64)
+        if (mbedtls_aesce_has_support()) {
+            mbedtls_printf("  GCM note: using AESCE.\n");
+        } else
```

Comment:
This won't compile on IAR (which does support AArch64) - you'll get `statement is unreachable` on the `else` clause with `MBEDTLS_AES_USE_HARDWARE_ONLY` - see #7932

Reply:
@tom-cosgrove-arm since Jerry is out at the moment, if this is your only comment, shall we merge and get Aggie to address the IAR issue separately? If you have other comments obviously Jerry can pick this up with the rest.

Reply:
Those won't raise `statement is unreachable` warning.  `mbedtls_padlock_has_support` and `mbedtls_aesni_has_support` are not defined as const value in some configure options.

Reply:
> @tom-cosgrove-arm since Jerry is out at the moment, if this is your only comment, shall we merge and get Aggie to address the IAR issue separately? If you have other comments obviously Jerry can pick this up with the rest.

I'd like resolve IAR issue separately. And I will limit the compiler to MSVC , GCC, CLANG and ArmClang in this PR.

Reply:
> are not defined as const value in some configure options

It's not about "some configure options", it's "with a given set of configure options": the compiler doesn't see all sets of pre-processor values, it only sees what gets through after the pre-processor has finished. So if we (or a user) compiles with IAR when `mbedtls_aesni_has_support()` is constant and true, it will complain that anything in the subsequent `else` is unreachable.

And note: I've seen warnings from some versions of Clang when the test in a if-statement is constants, too.

Reply:
I am removing the constant definition locally. 

> I've seen warnings from some versions of Clang when the test in a if-statement is constants, too.

Do you remember the version number? I do not have IAR for aarch64, I think I can use the Clang version to reproduce the issue.

Reply:
@tom-cosgrove-arm , I tried clang versions at https://godbolt.org/z/zraKEqqfG , but I could not reproduce the issue. Do you remember the version.

Reply:
I verified this issue with `clang -Wunreachable-code`

Reply:
Agree with Jerry, let's cover the main compilers in this PR and handle IAR separately - maybe @AgathiyanB can take a look after this has been merged